Implement JavaScript test coverage checking

unoduetre opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

Investigate ways of adding a JavaScript test coverage tool to our apps. Potential problems might involve the fact that some of our JavaScript tests run in a browser, so it might be difficult to make the test coverage tool aware of those tests.

If adding a test coverage tool turns out to be possible, discuss and implement the proposed solution.

## Why

In order to automatically check that the requirement of 95% test coverage is met, which is the requirement of https://github.com/alphagov/govuk-dependabot-merger which we use in our repositories to lessen the burden of the developers in relation to merging dependencies upgrades.

[Trello card](https://trello.com/c/6cX3Xipx/2664-investigate-adding-javascript-test-coverage-and-implement-it-if-possible-l), [Jira issue NAV-15267](https://gov-uk.atlassian.net/browse/NAV-15267)

## Dependencies

This PR depends on restructuring JavaScript files, so the code that should be checked for test coverage is separated from the third-party code e.g. components.

The PR where restructuring was done is https://github.com/alphagov/frontend/pull/4164

## Further info

The options of `.nycrc` are documented here: https://github.com/istanbuljs/nyc?tab=readme-ov-file#common-configuration-options

The following options have been set for the following reasons:
* `all=true` - instrument all JavaScript assets in the directory for storing test coverage;
* `exclude=...` - but not the dependencies, e.g. components that should be tested elsewhere;
* `branches/lines=95` - specify required test coverage (`lines` test coverage check is not strictly necessary because we check branches as well, but it doesn't hurt);
* `check-coverage=true` - returns unsuccessful exit code if test coverage requirements are not met (setting this to `false` would allow us to publish an urgent change in the unlikely scenario that we needed to override the code coverage check to do so; alternatively the `/* istanbul ignore next */` syntax could be used to exclude lines of code from the code coverage check) 
* `compact=false` - does not minify the resulting instrumented files;
* `report-dir=...` - specify where the html report should be stored;
* `reporter=...` - show text summary and create the html report;
* `source-map=false` - I had to disable that as it didn't seem to work with our source maps.

## Screenshots

### Text summary

<img width="1241" alt="Screenshot 2024-07-18 at 14 51 15" src="https://github.com/user-attachments/assets/7a65e2da-fd97-4533-9836-4d104a70e423">

### HTML report

![Screenshot 2024-07-18 at 14-53-11 Code coverage report ](https://github.com/user-attachments/assets/a936505f-c76d-432e-a396-4d5dfa575272)
![Screenshot 2024-07-18 at 14-53-34 Code coverage report ](https://github.com/user-attachments/assets/df40da3f-cde7-4e5f-9395-4e8a28bd6d66)

unoduetre committed:
Add JavaScript test coverage checking

unoduetre committed:
Add Jasmine GitHub workflow

unoduetre committed:
Increase JavaScript test coverage

unoduetre commented on app/assets/javascripts/views/travel-advice.js:
```diff
@@ -147,14 +148,11 @@
   }
 
-  CountryFilter.prototype.pagePath = function () {
-    window.location.pathname.split('/').pop()
```
This function is not used anywhere and it cannot work, because it doesn't do anything (the "return" is missing). It also doesn't have any tests.

unoduetre replied:
Removal of this function has been extracted to a separate PR.

unoduetre commented on config/application.rb:
```diff
@@ -48,12 +48,6 @@ class Application < Rails::Application
     # Enable the asset pipeline
     config.assets.enabled = true
 
-    config.assets.precompile += %w[
```
Moved to manifest.js

MartinJJones replied:
Just some history on modernizing the JS test approach in frontend app:
- https://github.com/alphagov/frontend/pull/1414 - 2018
added to jasmine.yml file
- https://github.com/alphagov/frontend/pull/3221 - 2022
jasmine.yml file removed, app.js not moved to jamine-browser.json

unoduetre replied:
This and related code has been extracted to a separate PR.

unoduetre commented on spec/support/reporters/test-coverage.mjs:
```diff
@@ -0,0 +1,23 @@
+    const nycOutputPath = path.join(rootDirname, '.nyc_output');
+    const assetsPath = path.join(nycOutputPath, 'assets');
+    const testCoveragePath = path.join(nycOutputPath, 'out.json');
+    return Promise.all([
```
Remove previous report and nyc-instrumented assets as they're not useful anymore and by removing them we make it more error-prone to run it multiple times, as they files related to the previous run are removed.

hannalaakso replied:
This seems like really useful information, should it be a code comment so that it gets retained for the next developer?

unoduetre replied:
Hmm, I wonder, if a code comment is the best idea. Could it be documented somewhere else? As this code will be shared between repos in the future, what would be the best place? Where would you search for this kind of info?

unoduetre replied:
*sorry, I meant **less** error-prone of course

hannalaakso replied:
As we discussed, having this as a code comment probably makes the most sense right now. Although it's possible that in the future we would have some designated place for docs once we share the functionality between repos.

Maybe we could shorten the code comment to say something like: 

```
// Remove the previous report and the nyc-instrumented assets as they are not needed after the run has completed.
```

unoduetre replied:
Done

unoduetre commented on app/assets/javascripts/all.js:
```diff
@@ -1,5 +0,0 @@
-// This is a manifest file used by the runner for JavaScript unit tests.
```
As confirmed by @MartinJJones, this file is not used anymore, so it can be removed.

MartinJJones replied:
Just some history on modernizing the JS test approach in frontend app:
- https://github.com/alphagov/frontend/pull/1414 - 2018
added to jasmine.yml file
- https://github.com/alphagov/frontend/pull/3221 - 2022
jasmine.yml file removed, app.js not moved to jamine-browser.json

unoduetre replied:
Removal of all.js has been extracted to a separate PR.

hannalaakso commented on spec/support/reporters/test-coverage.mjs:
```diff
@@ -8,11 +8,14 @@ export default class TestCoverageReporter {
-    const thisFileDirname = path.dirname(new URL(import.meta.url).pathname)
-    const nycOutputPath = path.join(thisFileDirname, '..', '..', '..', '.nyc_output');
+    const thisFileDirname = path.dirname(new URL(import.meta.url).pathname);
+    const rootDirname = path.join(thisFileDirname, '..', '..', '..');
```
I'm wondering if there's another way, maybe via Node, to get this root directory name. I'm just thinking that if this file gets moved, this path won't work anymore.

hannalaakso commented on spec/support/reporters/test-coverage.mjs:
```diff
@@ -8,11 +8,14 @@ export default class TestCoverageReporter {
-    const thisFileDirname = path.dirname(new URL(import.meta.url).pathname)
-    const nycOutputPath = path.join(thisFileDirname, '..', '..', '..', '.nyc_output');
+    const thisFileDirname = path.dirname(new URL(import.meta.url).pathname);
+    const rootDirname = path.join(thisFileDirname, '..', '..', '..');
```
I'm wondering if there's another way, maybe via Node, to get this root directory name. I'm just thinking that if this file gets moved, this path won't work anymore.

unoduetre replied:
Or alternatively we could use import.meta.resolve with a toplevel file.

unoduetre replied:
Please, check the discussion in one of the Martin's comments below.

I've also combined the "#root" commit with the main commit.

hannalaakso commented on spec/support/reporters/test-coverage.mjs:
Just a nit pick, but it looks like this commit could be squashed into the previous one as it's amending the new lines added in this PR.

unoduetre replied:
Done

MartinJJones commented on spec/support/jasmine-browser.json:
```diff
@@ -1,16 +1,20 @@
-  "browser": "headlessChrome"
+  "browser": "headlessChrome",
+  "reporters": [
+    "./spec/support/reporters/test-coverage.mjs"
```
Could we update the file name to include the name of the reporter, for example: `nyc-test-converage.js`

unoduetre replied:
Done

hannalaakso commented on spec/javascripts/helpers/test-coverage.mjs:
```diff
@@ -0,0 +1,9 @@
+window.__coverage__ ??= {};
+
+class TestCoverageReporter {
+  static suiteDone(result) {
```
Could we add a code comment here, something along the lines of:
`Runs after every top level Jasmine describe-block`

unoduetre replied:
Done

hannalaakso commented on spec/javascripts/helpers/test-coverage.mjs:
```diff
@@ -0,0 +1,9 @@
+
+class TestCoverageReporter {
+  static suiteDone(result) {
+    jasmine.getEnv().setSuiteProperty("__coverage__", window.__coverage__);
```
Could we add a code comment here eg. `Transfer the property from the browser to Node`

unoduetre replied:
Done.

hannalaakso commented on spec/javascripts/helpers/test-coverage.mjs:
```diff
@@ -0,0 +1,9 @@
+window.__coverage__ ??= {};
```
Could we add a code comment here at the top of the file to say:
`This file runs in the browser`

unoduetre replied:
Done.

hannalaakso commented on spec/support/reporters/test-coverage.mjs:
```diff
@@ -0,0 +1,24 @@
+import path from 'node:path';
+
+export default class TestCoverageReporter {
+  suiteDone({properties}) {
```
Could we add a code comment here, something along the lines of:
`Runs after every top level Jasmine describe-block`

unoduetre replied:
Done. I checked and it runs after every Jasmine block, not only top-level ones.

MartinJJones commented on spec/support/reporters/nyc-test-coverage-reporter.mjs:
```diff
@@ -0,0 +1,32 @@
+    const assetsPath = path.join(nycOutputPath, 'assets');
+    const testCoveragePath = path.join(nycOutputPath, 'out.json');
+    return Promise.all([
+      /* Remove the previous report and the nyc-instrumented assets
+       * as they are not needed after the run has completed,
+       * and if not removed, they would accumulate and potentially
+       * confuse the developer.
+       */
+      fsPromises.rm(reportPath, {recursive: true, force: true}),
+      fsPromises.rm(assetsPath, {recursive: true, force: true}),
```
After looking at this further we should be able to use the `delete` option in nyc to remove the output directory - 

> The --delete option will remove the existing output directory before instrumenting.

https://github.com/istanbuljs/nyc/blob/main/docs/instrument.md#pre-instrumented-source

For example, we would update this line - https://github.com/alphagov/frontend/pull/4142/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R12

From: `"nyc:instrument": "yarn run nyc instrument public/assets/frontend .nyc_output/assets"`

To: "nyc:instrument": "yarn run nyc instrument public/assets/frontend .nyc_output/assets --delete"

unoduetre replied:
A good idea, let me test it.

unoduetre replied:
Interestingly this options is not documented when one runs `yarn run nyc instrument --help` but it seems to work.

unoduetre replied:
Done

MartinJJones commented on .nycrc:
```diff
@@ -0,0 +1,22 @@
+  "report-dir": "coverage/javascript",
+  "reporter": [
+    "text",
+    "html"
```
I wonder if we should remove the `html` option from the base level config for now for the reasons below:

- It appears that the HTML report will also be created when running the Continuous Integration checks but is not required here
- We can always add this feature in the future for use in a development environment by creating a custom nyc dev config - https://github.com/istanbuljs/nyc?tab=readme-ov-file#adding-your-overrides
- Would allow us to remove some of the code used in the nyc-test-coverage-reporter file used to delete the reports - https://github.com/alphagov/frontend/pull/4142/files#diff-cad9e4a78533bc33cf9c9b72531808b7be8503ad13528c114c664067d243893cR27

unoduetre replied:
No, continuous integration runs jasmine:ci

Also, developers need html reports in order to check the coverage, otherwise it's difficult to discover what tests needs to be added.

MartinJJones replied:
> No, continuous integration runs jasmine:ci

`yarn run jasmine:prepare && yarn run jasmine-browser-runner runSpecs && yarn run nyc:report`

This script also appears to run `nyc:report`, at which point does the HTML report also get generated?

unoduetre replied:
Yes, but nyc:report only runs the reporters specified in this array, so no html report will be generated. You can check yourself by commenting `"html"`, removing the existing report and running `yarn run jasmine:ci` again.

unoduetre replied:
After the discussion we decided to add a separate package.json script for creating the html report and require the user to manually remove the previous report before generating a new one.

unoduetre replied:
Done

MartinJJones replied:
Thanks, I think we can always look to add another step / something in the build process to help improve the developer experience in the future to avoid manually deleting the files generated by the HTML report

MartinJJones commented on package.json:
```diff
@@ -8,10 +8,16 @@
+    "rails:precompile-assets": "RAILS_ENV=test bundle exec rails assets:clobber assets:precompile",
+    "nyc:instrument": "yarn run nyc instrument public/assets/frontend .nyc_output/assets",
+    "jasmine:prepare": "yarn run rails:precompile-assets && yarn run nyc:instrument",
+    "nyc:report": "yarn run nyc report",
```
Small nitpick: should we group similar scripts together? or put the scripts in alphabetical order?

Non-blocking as it feels more like something we should have a general consensus on and tooling in place to manage this, ES Lint for example - https://eslint.org/docs/latest/contribute/package-json-conventions#order

unoduetre replied:
Yeah, we can change the ordering after it's decided in another tiny PR.

MartinJJones commented on package.json:
```diff
@@ -8,10 +8,16 @@
     "jasmine:browser": "yarn run jasmine:prepare && yarn run jasmine-browser-runner"
   },
+  "imports": {
+    "#root/*": "./*"
```
Depending on if the other changes are made in [nyc-test-coverage-reporter.mjs](https://github.com/alphagov/frontend/pull/4142/files#diff-cad9e4a78533bc33cf9c9b72531808b7be8503ad13528c114c664067d243893c) we may only need a reference to nyc_output, although I can see the benefit of having a reference to the root directory as well, is it possible to add a reference to the nyc_output folder here as well?

Another approach that we could use is creating a JS file that contains the paths to be used in the application, example from the govuk-frontend (Design System)
https://github.com/alphagov/govuk-frontend/blob/a50eb60c9583aea38e82625889fa7917a2f2fad9/shared/config/paths.js

unoduetre replied:
We can also add a specific path instead of using a pattern.

unoduetre replied:
After the discussion we decided to keep the imports solution. Nevertheless, because of getting rid of file removal from the reporter, the reporter can be simplified.

unoduetre replied:
Done

MartinJJones commented on spec/support/reporters/nyc-test-coverage-reporter.mjs:
```diff
@@ -0,0 +1,32 @@
+
+  jasmineDone() {
+    const rootDirname = path.dirname(
+      new URL(import.meta.resolve('#root/package.json')).pathname
```
Looking at the documentation in Node it appears that `import.meta.resolve` is marked as experimental, [Stability: 1.2](https://nodejs.org/docs/latest/api/documentation.html#stability-index) - Release candidate. It appears the minimum node version required to use this is 18.20.4, using a version less than this produces the error `TypeError: (intermediate value).resolve is not a function`. Could there be any issues in using this?

unoduetre replied:
We control the local node version. `import.meta.resolve` is a part of html standard, not node, so there is no risk of this feature being removed or changed: https://html.spec.whatwg.org/multipage/webappapis.html#integration-with-the-javascript-module-system

MartinJJones replied:
I was thinking are not running it in a node context here? and if so does the statement from MDN apply:

> Node.js also implements [the import.meta.resolve() function](https://nodejs.org/docs/latest/api/esm.html#importmetaresolvespecifier), but adds an additional parent parameter if you use the --experimental-import-meta-resolve flag.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import.meta/resolve#not_an_ecmascript_feature

unoduetre replied:
This relates to an additional parameter to resolve() named "parent". We don't use this experimental feature in this code.

MartinJJones replied:
Thanks for clarifying 👍

MartinJJones commented:
Changes look good to me overall 👍, nice work, I just have some further suggestions / questions, nothing else to review from a frontend perspective after that, thanks 😃

unoduetre committed:
Use node 18.20.4

This is the latest LTS release from the 18 branch.
We decided to use the 18 branch as it matches the branch
in the GitHub CI configuration.

MartinJJones commented:
Great work on this!, happy to approve from a frontend perspective 👍

unoduetre commented:
> Great work on this!, happy to approve from a frontend perspective 👍

Thank you!

beccapearce reviewed (approved):
This all looks good to me!! 🚀 
nice one!
